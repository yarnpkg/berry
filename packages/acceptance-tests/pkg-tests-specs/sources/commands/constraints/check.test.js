const {
  fs: {createTemporaryFolder, mkdirp, writeFile, writeJson},
} = require('pkg-tests-core');

const environments = {
  [`empty project`]: async path => {
    await writeJson(`${path}/package.json`, {});
  },
  [`one regular dependency`]: async path => {
    await writeJson(`${path}/package.json`, {
      dependencies: {
        [`no-deps`]: `1.0.0`,
      }
    });
  },
  [`two regular dependencies`]: async path => {
    await writeJson(`${path}/package.json`, {
      dependencies: {
        [`no-deps`]: `1.0.0`,
        [`no-deps-bin`]: `1.0.0`,
      },
    });
  },
  [`two development dependencies`]: async path => {
    await writeJson(`${path}/package.json`, {
      devDependencies: {
        [`no-deps`]: `1.0.0`,
        [`no-deps-bin`]: `1.0.0`,
      },
    });
  },
  [`two regular dependencies, two development dependencies`]: async path => {
    await writeJson(`${path}/package.json`, {
      dependencies: {
        [`no-deps`]: `1.0.0`,
        [`no-deps-bin`]: `1.0.0`,
      },
      devDependencies: {
        [`no-deps`]: `1.0.0`,
        [`no-deps-bin`]: `1.0.0`,
      },
    });
  },
  [`multiple workspaces`]: async path => {
    await writeJson(`${path}/package.json`, {
      private: true,
      workspaces: [`packages/**`],
    });

    await mkdirp(`${path}/packages/workspace-a`);
    await mkdirp(`${path}/packages/workspace-b`);

    await writeJson(`${path}/packages/workspace-a/package.json`, {
      dependencies: {
        [`no-deps`]: `1.0.0`,
        [`no-deps-bin`]: `1.0.0`,
      },
      devDependencies: {
        [`no-deps`]: `1.0.0`,
        [`no-deps-bin`]: `1.0.0`,
      },
    });

    await writeJson(`${path}/packages/workspace-b/package.json`, {
      dependencies: {
        [`no-deps`]: `1.0.0`,
        [`no-deps-bin`]: `1.0.0`,
      },
      devDependencies: {
        [`no-deps`]: `1.0.0`,
        [`no-deps-bin`]: `1.0.0`,
      },
    });
  },
};

const constraints = {
  [`empty constraints`]: ``,
  [`gen_enforced_dependency_range (missing)`]: `gen_enforced_dependency_range(WorkspaceCwd, 'one-fixed-dep', '1.0.0', peerDependencies).`,
  [`gen_enforced_dependency_range (incompatible)`]: `gen_enforced_dependency_range(WorkspaceCwd, 'no-deps', '2.0.0', dependencies).`,
  [`gen_enforced_dependency_range (extraneous)`]: `gen_enforced_dependency_range(WorkspaceCwd, 'no-deps', null, _).`,
};

describe(`Commands`, () => {
  describe(`constraints check`, () => {
    for (const [environmentDescription, environment] of Object.entries(environments)) {
      for (const [scriptDescription, script] of Object.entries(constraints)) {
        // Note: Don't use concurrent testing w/ snapshots, they don't work together
        test(`test (${environmentDescription} / ${scriptDescription})`, makeTemporaryEnv({}, async ({path, run, source}) => {
          await environment(path);
          await writeFile(`${path}/constraints.pro`, script);

          let code;
          let stdout;
          let stderr;

          try {
            ({code, stdout, stderr} = await run(`constraints`, `check`));
          } catch (error) {
            ({code, stdout, stderr} = error);
          }

          expect({code, stdout, stderr}).toMatchSnapshot();
        }));
      }
    }
  });
});
