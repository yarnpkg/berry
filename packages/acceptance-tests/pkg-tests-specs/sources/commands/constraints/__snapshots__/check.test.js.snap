// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Commands constraints check test (empty project / empty constraints) 1`] = `
Object {
  "code": 0,
  "stderr": "",
  "stdout": "➤ YN0000: Done
",
}
`;

exports[`Commands constraints check test (empty project / gen_enforced_dependency_range (extraneous)) 1`] = `
Object {
  "code": 0,
  "stderr": "",
  "stdout": "➤ YN0000: Done
",
}
`;

exports[`Commands constraints check test (empty project / gen_enforced_dependency_range (incompatible)) 1`] = `
Object {
  "code": 1,
  "stderr": "",
  "stdout": "➤ YN0023: root-workspace-988eec must depend on no-deps (via 2.0.0) in dependencies, but doesn't
➤ YN0000: Failed with errors
",
}
`;

exports[`Commands constraints check test (empty project / gen_enforced_dependency_range (missing)) 1`] = `
Object {
  "code": 1,
  "stderr": "",
  "stdout": "➤ YN0023: root-workspace-988eec must depend on one-fixed-dep (via 1.0.0) in peerDependencies, but doesn't
➤ YN0000: Failed with errors
",
}
`;

exports[`Commands constraints check test (empty project / gen_invalid_dependency (deep field path)) 1`] = `
Object {
  "code": 0,
  "stderr": "",
  "stdout": "➤ YN0000: Done
",
}
`;

exports[`Commands constraints check test (empty project / gen_workspace_field_requirement (extraneous)) 1`] = `
Object {
  "code": 0,
  "stderr": "",
  "stdout": "➤ YN0000: Done
",
}
`;

exports[`Commands constraints check test (empty project / gen_workspace_field_requirement (incompatible)) 1`] = `
Object {
  "code": 1,
  "stderr": "",
  "stdout": "➤ YN0037: root-workspace-988eec must have field \\"dependencies[\\"no-deps\\"]\\" value \\"2.0.0\\", but doesn't
➤ YN0000: Failed with errors
",
}
`;

exports[`Commands constraints check test (empty project / gen_workspace_field_requirement (missing)) 1`] = `
Object {
  "code": 1,
  "stderr": "",
  "stdout": "➤ YN0037: root-workspace-988eec must have field \\"dependencies[\\"a-new-dep\\"]\\" value \\"1.0.0\\", but doesn't
➤ YN0000: Failed with errors
",
}
`;

exports[`Commands constraints check test (multiple workspaces / empty constraints) 1`] = `
Object {
  "code": 0,
  "stderr": "",
  "stdout": "➤ YN0000: Done
",
}
`;

exports[`Commands constraints check test (multiple workspaces / gen_enforced_dependency_range (extraneous)) 1`] = `
Object {
  "code": 1,
  "stderr": "",
  "stdout": "➤ YN0025: workspace-a-01beb5 has an extraneous dependency on no-deps in dependencies
➤ YN0025: workspace-a-01beb5 has an extraneous dependency on no-deps in devDependencies
➤ YN0025: workspace-b-d95f30 has an extraneous dependency on no-deps in dependencies
➤ YN0025: workspace-b-d95f30 has an extraneous dependency on no-deps in devDependencies
➤ YN0000: Failed with errors
",
}
`;

exports[`Commands constraints check test (multiple workspaces / gen_enforced_dependency_range (incompatible)) 1`] = `
Object {
  "code": 1,
  "stderr": "",
  "stdout": "➤ YN0023: root-workspace-988eec must depend on no-deps (via 2.0.0) in dependencies, but doesn't
➤ YN0024: workspace-a-01beb5 must depend on no-deps via 2.0.0 in dependencies, but uses 1.0.0 instead
➤ YN0024: workspace-b-d95f30 must depend on no-deps via 2.0.0 in dependencies, but uses 1.0.0 instead
➤ YN0000: Failed with errors
",
}
`;

exports[`Commands constraints check test (multiple workspaces / gen_enforced_dependency_range (missing)) 1`] = `
Object {
  "code": 1,
  "stderr": "",
  "stdout": "➤ YN0023: root-workspace-988eec must depend on one-fixed-dep (via 1.0.0) in peerDependencies, but doesn't
➤ YN0023: workspace-a-01beb5 must depend on one-fixed-dep (via 1.0.0) in peerDependencies, but doesn't
➤ YN0023: workspace-b-d95f30 must depend on one-fixed-dep (via 1.0.0) in peerDependencies, but doesn't
➤ YN0000: Failed with errors
",
}
`;

exports[`Commands constraints check test (multiple workspaces / gen_invalid_dependency (deep field path)) 1`] = `
Object {
  "code": 1,
  "stderr": "",
  "stdout": "➤ YN0026: workspace-a-01beb5 has an invalid dependency on no-deps in dependencies (invalid because no-deps is not allowed unless built is set to false)
➤ YN0026: workspace-a-01beb5 has an invalid dependency on no-deps in devDependencies (invalid because no-deps is not allowed unless built is set to false)
➤ YN0000: Failed with errors
",
}
`;

exports[`Commands constraints check test (multiple workspaces / gen_workspace_field_requirement (extraneous)) 1`] = `
Object {
  "code": 1,
  "stderr": "",
  "stdout": "➤ YN0039: workspace-a-01beb5 has an extraneous field \\"dependencies\\" with value {\\"no-deps\\":\\"1.0.0\\",\\"no-deps-bin\\":\\"1.0.0\\"}
➤ YN0039: workspace-b-d95f30 has an extraneous field \\"dependencies\\" with value {\\"no-deps\\":\\"1.0.0\\",\\"no-deps-bin\\":\\"1.0.0\\"}
➤ YN0000: Failed with errors
",
}
`;

exports[`Commands constraints check test (multiple workspaces / gen_workspace_field_requirement (incompatible)) 1`] = `
Object {
  "code": 1,
  "stderr": "",
  "stdout": "➤ YN0037: root-workspace-988eec must have field \\"dependencies[\\"no-deps\\"]\\" value \\"2.0.0\\", but doesn't
➤ YN0038: workspace-a-01beb5 must have field \\"dependencies[\\"no-deps\\"]\\" with value \\"2.0.0\\" but it has value \\"1.0.0\\"
➤ YN0038: workspace-b-d95f30 must have field \\"dependencies[\\"no-deps\\"]\\" with value \\"2.0.0\\" but it has value \\"1.0.0\\"
➤ YN0000: Failed with errors
",
}
`;

exports[`Commands constraints check test (multiple workspaces / gen_workspace_field_requirement (missing)) 1`] = `
Object {
  "code": 1,
  "stderr": "",
  "stdout": "➤ YN0037: root-workspace-988eec must have field \\"dependencies[\\"a-new-dep\\"]\\" value \\"1.0.0\\", but doesn't
➤ YN0037: workspace-a-01beb5 must have field \\"dependencies[\\"a-new-dep\\"]\\" value \\"1.0.0\\", but doesn't
➤ YN0037: workspace-b-d95f30 must have field \\"dependencies[\\"a-new-dep\\"]\\" value \\"1.0.0\\", but doesn't
➤ YN0000: Failed with errors
",
}
`;

exports[`Commands constraints check test (one regular dependency / empty constraints) 1`] = `
Object {
  "code": 0,
  "stderr": "",
  "stdout": "➤ YN0000: Done
",
}
`;

exports[`Commands constraints check test (one regular dependency / gen_enforced_dependency_range (extraneous)) 1`] = `
Object {
  "code": 1,
  "stderr": "",
  "stdout": "➤ YN0025: root-workspace-988eec has an extraneous dependency on no-deps in dependencies
➤ YN0000: Failed with errors
",
}
`;

exports[`Commands constraints check test (one regular dependency / gen_enforced_dependency_range (incompatible)) 1`] = `
Object {
  "code": 1,
  "stderr": "",
  "stdout": "➤ YN0024: root-workspace-988eec must depend on no-deps via 2.0.0 in dependencies, but uses 1.0.0 instead
➤ YN0000: Failed with errors
",
}
`;

exports[`Commands constraints check test (one regular dependency / gen_enforced_dependency_range (missing)) 1`] = `
Object {
  "code": 1,
  "stderr": "",
  "stdout": "➤ YN0023: root-workspace-988eec must depend on one-fixed-dep (via 1.0.0) in peerDependencies, but doesn't
➤ YN0000: Failed with errors
",
}
`;

exports[`Commands constraints check test (one regular dependency / gen_invalid_dependency (deep field path)) 1`] = `
Object {
  "code": 1,
  "stderr": "",
  "stdout": "➤ YN0026: root-workspace-988eec has an invalid dependency on no-deps in dependencies (invalid because no-deps is not allowed unless built is set to false)
➤ YN0000: Failed with errors
",
}
`;

exports[`Commands constraints check test (one regular dependency / gen_workspace_field_requirement (extraneous)) 1`] = `
Object {
  "code": 1,
  "stderr": "",
  "stdout": "➤ YN0039: root-workspace-988eec has an extraneous field \\"dependencies\\" with value {\\"no-deps\\":\\"1.0.0\\"}
➤ YN0000: Failed with errors
",
}
`;

exports[`Commands constraints check test (one regular dependency / gen_workspace_field_requirement (incompatible)) 1`] = `
Object {
  "code": 1,
  "stderr": "",
  "stdout": "➤ YN0038: root-workspace-988eec must have field \\"dependencies[\\"no-deps\\"]\\" with value \\"2.0.0\\" but it has value \\"1.0.0\\"
➤ YN0000: Failed with errors
",
}
`;

exports[`Commands constraints check test (one regular dependency / gen_workspace_field_requirement (missing)) 1`] = `
Object {
  "code": 1,
  "stderr": "",
  "stdout": "➤ YN0037: root-workspace-988eec must have field \\"dependencies[\\"a-new-dep\\"]\\" value \\"1.0.0\\", but doesn't
➤ YN0000: Failed with errors
",
}
`;

exports[`Commands constraints check test (two development dependencies / empty constraints) 1`] = `
Object {
  "code": 0,
  "stderr": "",
  "stdout": "➤ YN0000: Done
",
}
`;

exports[`Commands constraints check test (two development dependencies / gen_enforced_dependency_range (extraneous)) 1`] = `
Object {
  "code": 1,
  "stderr": "",
  "stdout": "➤ YN0025: root-workspace-988eec has an extraneous dependency on no-deps in devDependencies
➤ YN0000: Failed with errors
",
}
`;

exports[`Commands constraints check test (two development dependencies / gen_enforced_dependency_range (incompatible)) 1`] = `
Object {
  "code": 1,
  "stderr": "",
  "stdout": "➤ YN0023: root-workspace-988eec must depend on no-deps (via 2.0.0) in dependencies, but doesn't
➤ YN0000: Failed with errors
",
}
`;

exports[`Commands constraints check test (two development dependencies / gen_enforced_dependency_range (missing)) 1`] = `
Object {
  "code": 1,
  "stderr": "",
  "stdout": "➤ YN0023: root-workspace-988eec must depend on one-fixed-dep (via 1.0.0) in peerDependencies, but doesn't
➤ YN0000: Failed with errors
",
}
`;

exports[`Commands constraints check test (two development dependencies / gen_invalid_dependency (deep field path)) 1`] = `
Object {
  "code": 1,
  "stderr": "",
  "stdout": "➤ YN0026: root-workspace-988eec has an invalid dependency on no-deps in devDependencies (invalid because no-deps is not allowed unless built is set to false)
➤ YN0000: Failed with errors
",
}
`;

exports[`Commands constraints check test (two development dependencies / gen_workspace_field_requirement (extraneous)) 1`] = `
Object {
  "code": 0,
  "stderr": "",
  "stdout": "➤ YN0000: Done
",
}
`;

exports[`Commands constraints check test (two development dependencies / gen_workspace_field_requirement (incompatible)) 1`] = `
Object {
  "code": 1,
  "stderr": "",
  "stdout": "➤ YN0037: root-workspace-988eec must have field \\"dependencies[\\"no-deps\\"]\\" value \\"2.0.0\\", but doesn't
➤ YN0000: Failed with errors
",
}
`;

exports[`Commands constraints check test (two development dependencies / gen_workspace_field_requirement (missing)) 1`] = `
Object {
  "code": 1,
  "stderr": "",
  "stdout": "➤ YN0037: root-workspace-988eec must have field \\"dependencies[\\"a-new-dep\\"]\\" value \\"1.0.0\\", but doesn't
➤ YN0000: Failed with errors
",
}
`;

exports[`Commands constraints check test (two regular dependencies / empty constraints) 1`] = `
Object {
  "code": 0,
  "stderr": "",
  "stdout": "➤ YN0000: Done
",
}
`;

exports[`Commands constraints check test (two regular dependencies / gen_enforced_dependency_range (extraneous)) 1`] = `
Object {
  "code": 1,
  "stderr": "",
  "stdout": "➤ YN0025: root-workspace-988eec has an extraneous dependency on no-deps in dependencies
➤ YN0000: Failed with errors
",
}
`;

exports[`Commands constraints check test (two regular dependencies / gen_enforced_dependency_range (incompatible)) 1`] = `
Object {
  "code": 1,
  "stderr": "",
  "stdout": "➤ YN0024: root-workspace-988eec must depend on no-deps via 2.0.0 in dependencies, but uses 1.0.0 instead
➤ YN0000: Failed with errors
",
}
`;

exports[`Commands constraints check test (two regular dependencies / gen_enforced_dependency_range (missing)) 1`] = `
Object {
  "code": 1,
  "stderr": "",
  "stdout": "➤ YN0023: root-workspace-988eec must depend on one-fixed-dep (via 1.0.0) in peerDependencies, but doesn't
➤ YN0000: Failed with errors
",
}
`;

exports[`Commands constraints check test (two regular dependencies / gen_invalid_dependency (deep field path)) 1`] = `
Object {
  "code": 1,
  "stderr": "",
  "stdout": "➤ YN0026: root-workspace-988eec has an invalid dependency on no-deps in dependencies (invalid because no-deps is not allowed unless built is set to false)
➤ YN0000: Failed with errors
",
}
`;

exports[`Commands constraints check test (two regular dependencies / gen_workspace_field_requirement (extraneous)) 1`] = `
Object {
  "code": 1,
  "stderr": "",
  "stdout": "➤ YN0039: root-workspace-988eec has an extraneous field \\"dependencies\\" with value {\\"no-deps\\":\\"1.0.0\\",\\"no-deps-bin\\":\\"1.0.0\\"}
➤ YN0000: Failed with errors
",
}
`;

exports[`Commands constraints check test (two regular dependencies / gen_workspace_field_requirement (incompatible)) 1`] = `
Object {
  "code": 1,
  "stderr": "",
  "stdout": "➤ YN0038: root-workspace-988eec must have field \\"dependencies[\\"no-deps\\"]\\" with value \\"2.0.0\\" but it has value \\"1.0.0\\"
➤ YN0000: Failed with errors
",
}
`;

exports[`Commands constraints check test (two regular dependencies / gen_workspace_field_requirement (missing)) 1`] = `
Object {
  "code": 1,
  "stderr": "",
  "stdout": "➤ YN0037: root-workspace-988eec must have field \\"dependencies[\\"a-new-dep\\"]\\" value \\"1.0.0\\", but doesn't
➤ YN0000: Failed with errors
",
}
`;

exports[`Commands constraints check test (two regular dependencies, two development dependencies / empty constraints) 1`] = `
Object {
  "code": 0,
  "stderr": "",
  "stdout": "➤ YN0000: Done
",
}
`;

exports[`Commands constraints check test (two regular dependencies, two development dependencies / gen_enforced_dependency_range (extraneous)) 1`] = `
Object {
  "code": 1,
  "stderr": "",
  "stdout": "➤ YN0025: root-workspace-988eec has an extraneous dependency on no-deps in dependencies
➤ YN0025: root-workspace-988eec has an extraneous dependency on no-deps in devDependencies
➤ YN0000: Failed with errors
",
}
`;

exports[`Commands constraints check test (two regular dependencies, two development dependencies / gen_enforced_dependency_range (incompatible)) 1`] = `
Object {
  "code": 1,
  "stderr": "",
  "stdout": "➤ YN0024: root-workspace-988eec must depend on no-deps via 2.0.0 in dependencies, but uses 1.0.0 instead
➤ YN0000: Failed with errors
",
}
`;

exports[`Commands constraints check test (two regular dependencies, two development dependencies / gen_enforced_dependency_range (missing)) 1`] = `
Object {
  "code": 1,
  "stderr": "",
  "stdout": "➤ YN0023: root-workspace-988eec must depend on one-fixed-dep (via 1.0.0) in peerDependencies, but doesn't
➤ YN0000: Failed with errors
",
}
`;

exports[`Commands constraints check test (two regular dependencies, two development dependencies / gen_invalid_dependency (deep field path)) 1`] = `
Object {
  "code": 1,
  "stderr": "",
  "stdout": "➤ YN0026: root-workspace-988eec has an invalid dependency on no-deps in dependencies (invalid because no-deps is not allowed unless built is set to false)
➤ YN0026: root-workspace-988eec has an invalid dependency on no-deps in devDependencies (invalid because no-deps is not allowed unless built is set to false)
➤ YN0000: Failed with errors
",
}
`;

exports[`Commands constraints check test (two regular dependencies, two development dependencies / gen_workspace_field_requirement (extraneous)) 1`] = `
Object {
  "code": 1,
  "stderr": "",
  "stdout": "➤ YN0039: root-workspace-988eec has an extraneous field \\"dependencies\\" with value {\\"no-deps\\":\\"1.0.0\\",\\"no-deps-bin\\":\\"1.0.0\\"}
➤ YN0000: Failed with errors
",
}
`;

exports[`Commands constraints check test (two regular dependencies, two development dependencies / gen_workspace_field_requirement (incompatible)) 1`] = `
Object {
  "code": 1,
  "stderr": "",
  "stdout": "➤ YN0038: root-workspace-988eec must have field \\"dependencies[\\"no-deps\\"]\\" with value \\"2.0.0\\" but it has value \\"1.0.0\\"
➤ YN0000: Failed with errors
",
}
`;

exports[`Commands constraints check test (two regular dependencies, two development dependencies / gen_workspace_field_requirement (missing)) 1`] = `
Object {
  "code": 1,
  "stderr": "",
  "stdout": "➤ YN0037: root-workspace-988eec must have field \\"dependencies[\\"a-new-dep\\"]\\" value \\"1.0.0\\", but doesn't
➤ YN0000: Failed with errors
",
}
`;
